File: tessoku/c++/section3/a12.cpp
#include <bits/stdc++.h>

using namespace std;

int main() {

    /* 注意：10^9はlong longじゃないと入らない */

    long long N, K;

    long long As[100009];

    cin >> N >> K;

    for (int i = 1; i <= N; i++) {
        cin >> As[i];
    }

    int left = 1;
    long long right = 1000000000;
    long long mid;

    // 二分探索
    while (left <= right) {

        mid = (left + right) / 2;

        // midの時の印刷枚数を計算
        long long X = 0;

        for (int i = 1; i <= N; i++) {
            X += floor(mid / As[i]);
        }

        // デバッグ
        // cout << "left: " << left << " right: " << right << " mid: " << mid << " X: " << X << endl;

        if (X < K) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
 
    }

    cout << left << endl;

}

File: tessoku/python/section3/a12.py
N, K = map(int, input().split())

As = list(map(int, input().split()))


left = 0
right = 10**9 - 1

while left <= right:

    mid = (left + right) // 2

    # midの時の印刷枚数を計算
    X = 0
    for i in range(N):
        X += int(mid / As[i])
    
    # デバッグ
    # print('left:', left, 'right:', right, 'mid:', mid, 'X:', X)
    
    if X < K:
        left = mid + 1
    else:
        right = mid - 1
        
print(left)